### PR TITLE
Notebook uploader

### PR DESCRIPTION
**Pull request by johnkerl**
Ready for review, with the following caveat:

* We might decide how much of the auto-increment logic to port (or not to port) from `TileDB-Cloud-Jupyter-Contents`.
*  If destination exists, three options:
    * Reject the write with 'already exists'. 
    * Auto-increment (as done in Jupyter)
    * Overwrite current version (should not be the default; perhaps should be supported as an option)
* Today's code does the first option and can be merged if everyone's OK with that for now.

# Manual test scenario

* Use `download-notebook.py` (below) to get valid JSON
* Use `upload-notebooy.py`
* Click through in the TileDB Cloud console and verify all tabs work for the newly created notebook

```
$ cat download-notebook.py
#!/usr/bin/env python

# ================================================================
# Setup:
# * https://console.dev.tiledb.io
# * Set my storage path and storage credential name in Settings
#
# Note: cannot be invoked from within the base directory of either the
# TileDB-Py or TileDB-Cloud-Py repos -- or any directory with a subdirectory
# named `tiledb` -- since `import tiledb` / `import tiledb.cloud` clashes
# with the existence of that subdirectory.
# ================================================================

import sys
import tiledb.cloud

config = tiledb.Config()
config["rest.username"] = "johnkerl"
config["rest.password"] = "redacted"
# API endpoint used by https://console.dev.tiledb.io:
config["rest.server_address"] = "api.dev.tiledb.io"

# This is the array-URI format in TileDB Cloud:
tiledb_uri              = "tiledb://seth/Voila-Covid-Demo"
storage_credential_name = "my-sandbox-creds"
ipynb_file_name         = "/Users/johnkerl/Desktop/PLAN/foo.ipynb"

if len(sys.argv) == 3:
    tiledb_uri      = sys.argv[1]
    ipynb_file_name = sys.argv[2]

tiledb.cloud.download_ipnyb_file_name_from_cloud(
    tiledb_uri,
    storage_credential_name,
    ipynb_file_name,
)

print("Wrote "+ipynb_file_name)
```

```
$ cat upload-notebook.py
#!/usr/bin/env python

# ================================================================
# Setup:
# * https://console.dev.tiledb.io
# * Set my storage path and storage credential name in Settings
#
# Note: cannot be invoked from within the base directory of either the
# TileDB-Py or TileDB-Cloud-Py repos -- or any directory with a subdirectory
# named `tiledb` -- since `import tiledb` / `import tiledb.cloud` clashes
# with the existence of that subdirectory.
# ================================================================

import sys
import tiledb.cloud

tiledb.cloud.login(
    host="https://api.dev.tiledb.io/v1",
    username="johnkerl",
    password="redacted",
)

ipynb_file_name         = "/Users/johnkerl/Desktop/PLAN/foo.ipynb"
namespace               = "johnkerl"
array_name              = "testing-upload-014"
storage_path            = "s3://tiledb-johnkerl"
storage_credential_name = "my-sandbox-creds"

if len(sys.argv) == 3:
    ipynb_file_name = sys.argv[1]
    array_name      = sys.argv[2]

tiledb_uri = tiledb.cloud.upload_ipnyb_file_name_to_cloud(
    ipynb_file_name,
    namespace,
    array_name,
    storage_path,
    storage_credential_name,
)

print("tiledb_uri="+tiledb_uri)
```

Be sure to run this for the cases when the destination array does not already exist, and does already exist.

# Tracking

CH 7290

### CODE REVIEW

File: tiledb/cloud/__init__.py
Comment (johnkerl) [inline tiledb/cloud/__init__.py:24]: Nothing needed here to export `tiledb.cloud.notebook.upload_ipnyb_file_name_to_cloud`. These are to let people access `upload_ipnyb_file_name_to_cloud` from `import tiledb.cloud` rather than `import tiledb.cloud.notebook`.

File: tiledb/cloud/notebook.py
Comment (thetorpedodog) [inline tiledb/cloud/notebook.py:102]: This should probably be `tobytes()`, since `tostring()` is a deprecated alias.

I also don’t think we want a special error handler here; if the data we get back is invalid UTF-8, that is a problem that we probably don’t want to paper over.

File: tiledb/cloud/notebook.py
Comment (thetorpedodog) [inline tiledb/cloud/notebook.py:69]: I am open to correction from other reviewers but, for easier reading as a user, I prefer to wrap the docstrings at 80 cols and continue the descriptions on the next line with a two-space indent:

```
    :param storage_credential_name: yada yada yada
      yada yada
```

File: tiledb/cloud/notebook.py
Comment (thetorpedodog) [inline tiledb/cloud/notebook.py:304]: This code is unreachable.

File: tiledb/cloud/notebook.py
Comment (thetorpedodog) [inline tiledb/cloud/notebook.py:312]: This probably shouldn't be here anyway (no point in having a catch clause that immediately re-raises and does nothing else) but in the case where you *do* want to re-raise an exception (e.g. after filtering something out), you should use the bare `raise`, which will bubble the exception up without rewriting its call stack etc.

**Review by thetorpedodog — COMMENTED** [review]
lots of little formatting things as well (mixes of `"` and `'`, stray spaces, etc.) that Black should fix automatically too

File: tiledb/cloud/notebook.py
Comment (johnkerl) [reply tiledb/cloud/notebook.py:19]: vim + my conscience, LOL -- thanks! :)

File: tiledb/cloud/notebook.py
Comment (johnkerl) [reply tiledb/cloud/notebook.py:64]: awesome py-fu, thank you! :)

File: tiledb/cloud/__init__.py
Comment (johnkerl) [reply tiledb/cloud/__init__.py:24]: thanks! :)

**Comment by johnkerl** [comment]
@thetorpedodog thanks for all the great feedback -- ready for round two! :)

File: tiledb/cloud/notebook.py
Comment (johnkerl) [reply tiledb/cloud/notebook.py:63]: My typo in the docstring; should have been `tiledb://TileDB-Inc/quickstart_dense`. (Fixed.)

File: tiledb/cloud/notebook.py
Comment (johnkerl) [reply tiledb/cloud/notebook.py:97]: good catch, thanks!

File: tiledb/cloud/notebook.py
Comment (johnkerl) [reply tiledb/cloud/notebook.py]: i defer to your style preference ;) -- i do like if/else; i find them more clear to the reader on a visual scan, even if not needed by the machine.

File: tiledb/cloud/notebook.py
Comment (johnkerl) [reply tiledb/cloud/notebook.py:298]: thanks @thetorpedodog ! this was a holdover from the idea of implementing auto-increment (which apparently we aren't interested in doing on this PR). here i was thinking of separating out different ways of failing. but without the auto-increment, good call, this is just a needless remnant of an unimplemented idea.

**Review by johnkerl — COMMENTED** [review]
@thetorpedodog ready for round three! :)

File: tiledb/cloud/notebook.py
Comment (thetorpedodog) [inline tiledb/cloud/notebook.py:33]: `str` no longer needed since that information appears in the signature

File: tiledb/cloud/notebook.py
Comment (thetorpedodog) [inline tiledb/cloud/notebook.py:183]: weird that black gave you this. will it let you do

```python
ctx = tiledb.cloud.Ctx(
    {"rest.creation_access_credentials_name": storage_credential_name},
)
```
?

File: tiledb/cloud/notebook.py
Comment (thetorpedodog) [inline tiledb/cloud/notebook.py:257]: recommend throwing an `!r` in these formats so that the user gets nice quotation marks.

File: tiledb/cloud/notebook.py
Comment (thetorpedodog) [inline tiledb/cloud/notebook.py:259]: I would bump this in by four spaces so that it is in the except block and note that it is "retry other TileDBErrors", because other exception types will still bubble up.

File: tiledb/cloud/notebook.py
Comment (thetorpedodog) [inline tiledb/cloud/notebook.py:342]: no need for `return` at the end of a void function.

File: tiledb/cloud/notebook.py
Comment (thetorpedodog) [inline tiledb/cloud/notebook.py:294]: Just use `tiledb.Array.create` here; in nearly all cases you should never have to directly reference `DenseArray` or `SparseArray` directly. It knows which one to create based on whether the schema has `sparse=False` or `sparse=True`.

File: tiledb/cloud/notebook.py
Comment (thetorpedodog) [inline tiledb/cloud/notebook.py]: `isort .`

File: tiledb/cloud/notebook.py
Comment (thetorpedodog) [inline tiledb/cloud/notebook.py:300]: Since you've `from tiledb.cloud import array`, you should use `array.update_info` here.

File: tiledb/cloud/notebook.py
Comment (thetorpedodog) [inline tiledb/cloud/notebook.py:261]: It would probably be best to let the last exception bubble out, probably by moving the final exception check inside the loop `except`:

```python
try:
  # ...
except TileDBException as e:
  if ...:
    ...
  if retries == 0:
    raise
```

On that last `raise`, we may specifically want to do the `check_exc` stuff that we do elsewhere in the codebase. I am not *super* familiar with the semantics or reason behind that so I have to defer to another reviewer for that.

**Review by thetorpedodog — COMMENTED** [review]
Nothing major left; just a few nits. Some of these comments may be out of date because as I was reviewing you pushed a new version which fixed some of them, but I have no idea how to find the old ones and get rid of them.

**Comment by thetorpedodog** [comment]
Also, annoyingly, it looks like your test failures are due to 502 errors :(

File: tiledb/cloud/notebook.py
Comment (johnkerl) [reply tiledb/cloud/notebook.py:33]: this is not a function i touched -- it was the function that was here when i started. and its doc string is out of data. and that info does not appear in the array signature.

i'll (try to) clean this up while i'm here, even though i didn't introduce it.

File: tiledb/cloud/notebook.py
Comment (johnkerl) [reply tiledb/cloud/notebook.py:183]: yup, black is good with either

File: tiledb/cloud/notebook.py
Comment (johnkerl) [reply tiledb/cloud/notebook.py:257]: TIL! :)

File: tiledb/cloud/notebook.py
Comment (johnkerl) [reply tiledb/cloud/notebook.py]: i have done so many times and i am now doing so on every commit.

* what formatting would you prefer?
* does isort give you something different?

File: tiledb/cloud/notebook.py
Comment (johnkerl) [reply tiledb/cloud/notebook.py:294]: TIL! :)

File: tiledb/cloud/notebook.py
Comment (thetorpedodog) [reply tiledb/cloud/notebook.py]: This was one of those comments where one of your follow-ons ended up fixing the problem and the comment became outdated; what appears in efdc478a30413c37d3ba466eab77c64621e497c4 is The Correct Format.

File: tiledb/cloud/notebook.py
Comment (johnkerl) [reply tiledb/cloud/notebook.py:261]: i prefer to do that on a follow-on PR -- i've added that to ch7290 -- thanks!

File: tiledb/cloud/notebook.py
Comment (thetorpedodog) [inline tiledb/cloud/notebook.py:249]:
```suggestion
                raise tiledb_cloud_error.TileDBCloudError(
```

(likewise below; this is not an appropriate situation for a `ValueError`; the value is perfectly valid, but it does not work given the environment. for comparison, `os.mkdir("")` might be a `ValueError`, since making an empty directory makes no sense, but `os.mkdir("/path/to/somewhere/that/already/exists")` is *not* a ValueError.)

**Review by thetorpedodog — APPROVED** [review]
This looks good! In a follow-up it might be good to add an automated test where we upload a notebook from a file, then download it to a different file, and perform a diff. That will also require a call to find the current user’s default storage path (since, ideally, we do not want to hardcode in our own unit test user’s default path).

If you want to wait for comments from others you may do so but (apart from my suggestion below) this is good to go imo.

In the interest of keeping an easier-to-understand change history I would recommend the “squash and merge” option when submitting.

File: tiledb/cloud/notebook.py
Comment (johnkerl) [reply tiledb/cloud/notebook.py:249]: > In the interest of keeping an easier-to-understand change history I would recommend the “squash and merge” option when submitting.

100% my preference as well @thetorpedodog -- thanks! :)

**Pull request merged**
